src/core/processing/qgsprocessingalgorithm.cpp

```diff
@@ -608,3 +608,67 @@ bool QgsProcessingAlgorithm::createAutoOutputForParameter( QgsProcessingParamete
 }
 
 
+//
+// QgsProcessingFeatureBasedAlgorithm
+//
+
+void QgsProcessingFeatureBasedAlgorithm::initAlgorithm( const QVariantMap &config )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ) ) );
+  initParameters( config );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), outputName(), outputLayerType() ) );
+}
+
+QgsCoordinateReferenceSystem QgsProcessingFeatureBasedAlgorithm::sourceCrs() const
+{
+  if ( mSource )
+    return mSource->sourceCrs();
+  else
+    return QgsCoordinateReferenceSystem();
+}
+
+QVariantMap QgsProcessingFeatureBasedAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  mSource.reset( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !mSource )
+    return QVariantMap();
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest,
+                                          outputFields( mSource->fields() ),
+                                          outputWkbType( mSource->wkbType() ),
+                                          outputCrs( mSource->sourceCrs() ) ) );
+  if ( !sink )
+    return QVariantMap();
+
+  long count = mSource->featureCount();
+  if ( count <= 0 )
+    return QVariantMap();
```

Comment:
How does that behave in a model with an empty layer? Will the next algorithm not complain that it wasn't provided with a layer (while IMO the expected behavior is that it gets provided with an empty layer)?